No name bug when saving a role

AliRizvi1 opened the pull request:
https://dimagi-dev.atlassian.net/browse/SAAS-10974

##### SUMMARY
Added 'allowSubmit' function to roles.js, allowing it to automatically disable the 'Save' button if there was no name provided. Also changed the submission to a submit button in order to function properly with the 'allowSubmit' functionality.

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### RISK ASSESSMENT / QA PLAN
<!-- Does this need QA before or after merge? Link QA ticket. -->

##### PRODUCT DESCRIPTION
Users can no longer create a new role with no name, nor can they edit an existing role to have no name.

AliRizvi1 committed:
Added allowSubmit functionality, not yet connected to the button.

AliRizvi1 committed:
Fixed - users now have to input a role now before submitting

AliRizvi1 committed:
Clean up empty spaces and unused code

esoergel commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -192,9 +193,37 @@ hqDefine('users/js/roles',[
+        });
+
+        self.submitNewRole = function () {
+            // moved saveOptions inline
```
can you also delete the whole `modalSaveButton`, since it's no longer in-use?

esoergel commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -192,9 +193,37 @@ hqDefine('users/js/roles',[
         return self;
     };
 
+    //function(ko) return:
```
```suggestion
```

esoergel commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -20,13 +20,14 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
                      id="role-name"
-                     class="form-control" />
+                     class="form-control"
+                     required />
```
Ah, I think I see why this wasn't working - it isn't in the `<form />` block.  Can you see what happens if you extend the form element to encompass all of these fields?  (Conceptually, they _are_ all part of the form)

AliRizvi1 replied:
I just tried that, and it didn't seem to help (never highlighted the input box as red). I think this is because of the issue we discussed before, where the `enable: allowSubmit` means that the button is unable to submit unless there is an input, rendering the `required` flag unnecessary.
The mobile workers page works the same way, where they just added an asterisk next to the required inputs but otherwise don't highlight to the user what inputs are needed.

esoergel replied:
Did you remove the `novalidate` when you tried that?  I just mocked it up locally, and this is what I see when I click "Save" with the name field empty:
![image](https://user-images.githubusercontent.com/2367539/84941979-2897ef80-b0b0-11ea-9f8d-a68cd28ae5b2.png)

AliRizvi1 replied:
Ah no I forgot about the `novalidate`. I'll try it again.

esoergel replied:
Hopefully that should make it obvious to the user what the issue is, and it will probably make `allowSubmit` unnecessary.

AliRizvi1 replied:
In your local environment, you don't have the same `enable: allowSubmit` on the save button, right? Because even after creating an all-encompassing form and removing `novalidate` I still can't click the button when there is no name input.

I'm not sure how the `required` approach would fit in alongside the current `allowSubmit` functionality.

esoergel commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -715,23 +716,28 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
-          </button>
-        {% endif %}
-      </div>
+      <form novalidate data-bind="submit: $root.submitNewRole">
```
what does the `novalidate` do here?

AliRizvi1 replied:
It was used here: https://github.com/dimagi/commcare-hq/blob/591ede5b97ec68c6dca2c6db6706beda926c1039/corehq/apps/users/templates/users/mobile_workers.html#L88

So I just copied that. It seems to prevent native validation of form elements.

orangejenny commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -1,5 +1,6 @@
 hqDefine('users/js/roles',[
     'knockout',
+    'jquery',
```
You need to also add `$` to the parameters list. It works as is because we're currently in a weird state thats' half-migrated to requirejs, but the ideal future state is all dependencies are explicitly declared.

orangejenny commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -192,9 +193,37 @@ hqDefine('users/js/roles',[
+                    self.unsetRoleBeingEdited();
+                },
+                error: function (data) {
+                    console.log('error');
```
It'd be helpful for this to show the user there was a problem. One easy thing to do is use the `alert_user` library, which will display a message at the top of the page. One example: https://github.com/dimagi/commcare-hq/blob/d2da78ecdb8660ab5a185c413ead2060362f4b86/corehq/apps/case_search/static/case_search/js/case_search.js#L62-L64

orangejenny commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -20,13 +20,14 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
             <label class="col-sm-4 control-label"
                    for="role-name">
-              {% trans "Role Name" %}
+              {% trans "Role Name*" %}
```
Adding the asterisk is a nice touch. This is the nitpickiest of nitpicks, but can you wrap it in `<span class="asteriskField">`?

It's a super-nitpick because that class doesn't currently add any styles. But if we do want to style these asterisks, it's going to be ~impossible to find them, so it's good to mark them with a class.

orangejenny commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -192,9 +193,37 @@ hqDefine('users/js/roles',[
+            return true;
+        });
+
+        self.submitNewRole = function () {
```
Don't you need to return false to keep this from reloading the page, now that you're genuinely submitting a form?

AliRizvi1 replied:
Ethan showed me this: https://knockoutjs.com/documentation/submit-binding.html

So it seems like with knockout, it's not necessary to explicitly declare `return false`.

orangejenny replied:
nice, thanks for explaining

orangejenny commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -192,9 +193,37 @@ hqDefine('users/js/roles',[
+                dataType: 'json',
+                success: function (data) {
+                    self.addOrReplaceRole(data);
+                    self.unsetRoleBeingEdited();
```
Out of curiosity, is this what closes the modal?

AliRizvi1 replied:
To be honest I'm not entirely sure what's closing the modal, although that seems like where it's coming from.

orangejenny reviewed (commented):
Looks good overall, just had some minor comments.

AliRizvi1 committed:
Minor changes based on pull request comments

esoergel commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -18,10 +18,12 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
-                   for="role-name">
-              {% trans "Role Name*" %}
-            </label>
+            <span class="asteriskField">
```
I think Jenny meant for this `span` to encompass only the asterisk, eg:
https://github.com/dimagi/commcare-hq/blob/1cbafaaf72d864a295fb0317b000b44a39c522e2/corehq/motech/dhis2/templates/dhis2/partials/dhis2_dataset_map_templates.html#L17-L17

AliRizvi1 committed:
Made the role name a required field, removed previous allowSubmit functionality

AliRizvi1 commented:
The reason the last commit shows as 'large changes' (in `edit_role_modal.html`) is because I added a new form, and corrected the spacing to correspond with this form. GitHub recognizes that as me changing every line, however.

esoergel commented:
You can tell github to ignore whitespace changes, which will make the diff appear more readable.  
![image](https://user-images.githubusercontent.com/2367539/84947875-fa6add80-b0b8-11ea-8dc8-d2205450ba38.png)

AliRizvi1 committed:
Merged conflicts w/ master branch
